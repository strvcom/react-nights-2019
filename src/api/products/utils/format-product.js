export const formatProduct = (data, included) => ({
  id: data.id,
  ...data.attributes,
  price: included.find(
    price => price.id === data.relationships.prices.data[0].id
  ).attributes,
})
