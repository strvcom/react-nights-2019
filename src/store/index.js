import { createStore, combineReducers } from 'redux'

import products from './products'
import cartItems from './cartItems'

const reducer = combineReducers({
  products,
  cartItems,
})

const store = createStore(
  reducer,
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
)

export default store
