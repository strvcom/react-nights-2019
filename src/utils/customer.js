export const getCustomer = () => {
  const customer = window.localStorage.getItem('customer')
  if (customer) {
    return JSON.parse(customer)
  }
  return {}
}

export const setCustomer = customer => {
  window.localStorage.setItem('customer', JSON.stringify(customer))
}

export const removeCustomer = () => {
  window.localStorage.removeItem('customer')
}
