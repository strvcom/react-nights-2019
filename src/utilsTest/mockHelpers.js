import fetchMock from 'fetch-mock'
import config from '../config'
import {
  PRODUCTS,
  PRODUCT,
  USER,
  LOGIN_DATA,
  TOKEN,
  ERROR_EMAIL_ALREADY_TAKEN,
} from './mockData'

export const mockUnauthorizedRequest = () =>
  fetchMock.restore().post(`${config.apiUrl}/oauth/token`, {
    status: 401,
    body: {},
  })

export const mockAuthorizedRequest = () =>
  fetchMock
    .restore()
    .post(`${config.apiUrl}/oauth/token`, {
      status: 200,
      body: TOKEN,
    })
    .get(`${config.apiUrl}/api/customers/${USER.id}`, {
      status: 200,
      body: {
        data: LOGIN_DATA,
      },
    })

export const mockFetchProducts = () =>
  mockUnauthorizedRequest().get(
    `${
      config.apiUrl
    }/api/skus?include=prices&page%5Bnumber%5D=1&page%5Bsize%5D=50`,
    {
      status: 200,
      body: PRODUCTS,
    }
  )

export const mockFetchProduct = () =>
  mockUnauthorizedRequest().get(`${config.apiUrl}/api/skus/1?include=prices`, {
    status: 200,
    body: PRODUCT,
  })

export const mockSignUp422 = () =>
  mockUnauthorizedRequest().post(`${config.apiUrl}/api/customers`, {
    status: 411,
    body: {
      errors: [ERROR_EMAIL_ALREADY_TAKEN],
    },
  })

export const mockSignUp201 = () =>
  mockAuthorizedRequest().post(`${config.apiUrl}/api/customers`, {
    status: 201,
    body: {
      data: LOGIN_DATA,
    },
  })
