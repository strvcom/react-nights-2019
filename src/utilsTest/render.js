import React from 'react'
import { MemoryRouter } from 'react-router-dom'
import { render } from 'react-testing-library'

export const renderWithRouter = (component, uri) =>
  render(<MemoryRouter initialEntries={[uri]}>{component}</MemoryRouter>)
