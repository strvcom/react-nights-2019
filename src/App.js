import React from 'react'
import { Switch, Route, Redirect } from 'react-router-dom'
import { Provider } from 'react-redux'

import GlobalStyles from './globalStyles'
import { ProductList } from './pages/ProductList'
import { ProductDetail } from './pages/ProductDetail'
import { Cart } from './pages/Cart'
import { SignUp } from './pages/SignUp'
import { LogIn } from './pages/LogIn'
import { Account } from './pages/Account'
import { NotFound } from './pages/NotFound'
import { PrivateRoute } from './components/PrivateRoute'
import { getCustomer } from './utils/customer'
import { configureStore } from './store'
import * as routes from './routes'

const store = configureStore({
  customer: getCustomer(),
})

const App = ({ customStore }) => (
  <Provider store={customStore || store}>
    <React.Fragment>
      <GlobalStyles />
      <Switch>
        <Route
          path={routes.HOMEPAGE}
          exact
          render={() => <Redirect to={routes.PRODUCT_LIST} />}
        />
        <Route path={routes.PRODUCT_LIST} exact component={ProductList} />
        <Route path={routes.PRODUCT_DETAIL} component={ProductDetail} />
        <Route path={routes.CART} component={Cart} />
        <Route path={routes.SIGN_UP} component={SignUp} />
        <Route path={routes.LOGIN} component={LogIn} />
        <PrivateRoute path={routes.ACCOUNT} component={Account} />
        <Route component={NotFound} />
      </Switch>
    </React.Fragment>
  </Provider>
)

export { App }
