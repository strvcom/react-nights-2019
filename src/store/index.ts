import { createStore, combineReducers, applyMiddleware, compose } from 'redux'
import thunk from 'redux-thunk'

import { getCustomer } from '../utils/customer'
import { isBrowser } from '../utils/is-browser'

import products from './products/reducer'
import cart from './cart/reducer'
import customer from './customer/reducer'

const reducer = combineReducers({
  products,
  cart,
  customer,
})

export type AppState = ReturnType<typeof reducer>

// this variable will be set if you have redux-dev-tools extension installed in your browser
// https://github.com/zalmoxisus/redux-devtools-extension#12-advanced-store-setup
const composeEnhancers =
  // @ts-ignore
  (isBrowser() && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose

export const configureStore = (preloadedState = {}) =>
  createStore(
    reducer,
    { ...preloadedState, customer: getCustomer() },
    composeEnhancers(applyMiddleware(thunk))
  )
