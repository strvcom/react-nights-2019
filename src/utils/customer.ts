import { CustomerType } from '../common/types'
import { isBrowser } from './is-browser'
import { CustomerState } from '../store/customer/reducer'

export const getCustomer = (): CustomerState => {
  const customer = isBrowser() && window.localStorage.getItem('customer')
  if (customer) {
    try {
      return JSON.parse(customer)
    } catch (err) {
      console.log(err)
      return null
    }
  }
  return null
}

export const setCustomer = (customer: CustomerType) => {
  isBrowser() &&
    window.localStorage.setItem('customer', JSON.stringify(customer))
}

export const removeCustomer = () => {
  isBrowser() && window.localStorage.removeItem('customer')
}
