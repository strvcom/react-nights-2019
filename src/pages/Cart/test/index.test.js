import React from 'react'
import { waitForElement } from 'react-testing-library'
import 'jest-styled-components'

import { App } from '../../../App'
import { renderWithRouter } from '../../../utilsTest/render'
import { mockFetchProduct } from '../../../utilsTest/mockHelpers'
import { configureStore } from '../../../store'
import { getCustomer } from '../../../utils/customer'

describe('[pages] Cart', () => {
  describe('when loading', () => {
    it('should render correctly', () => {
      const renderer = renderWithRouter(<App />, '/cart')
      expect(renderer.baseElement).toMatchSnapshot()
    })
  })

  describe('when products loaded', () => {
    mockFetchProduct()

    it('should render correctly', async () => {
      const productId = 1
      const store = configureStore({
        customer: getCustomer(),
        cart: {
          [productId]: productId,
        },
      })

      const renderer = renderWithRouter(<App customStore={store} />, '/cart')
      await waitForElement(() => renderer.container.querySelector('p'))
      expect(renderer.baseElement).toMatchSnapshot()
    })
  })
})
