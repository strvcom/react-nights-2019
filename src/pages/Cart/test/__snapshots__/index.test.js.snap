// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[pages] Cart when loading should render correctly 1`] = `
.c2 {
  padding: 2rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-bottom: 0.1rem solid gainsboro;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 3rem;
}

.c1 {
  margin: 0 1rem;
}

.c3 {
  font-family: circular,serif;
  font-weight: 400;
  color: #ef0d33;
  text-transform: uppercase;
}

<body>
  <div>
    <header
      class="c0"
    >
      <div
        class=""
      >
        <a
          class="c1"
          href="/products"
        >
          All Products
        </a>
      </div>
      <div
        class=""
      >
        <a
          class="c1"
          href="/cart"
        >
          My Cart
        </a>
        |
        <a
          class="c1"
          href="/login"
        >
          Log In
        </a>
         |
        <a
          class="c1"
          href="/signup"
        >
          Sign Up
        </a>
      </div>
    </header>
    <div
      class="c2"
    >
      <h1
        class="c3"
      >
        Your cart
      </h1>
      <ul />
    </div>
  </div>
</body>
`;

exports[`[pages] Cart when products loaded should render correctly 1`] = `
<body>
  <div>
    <header
      class="sc-bwzfXH kuSkvH"
    >
      <div
        class="sc-htpNat KtFsv"
      >
        <a
          class="sc-bxivhb cHbCFb"
          href="/products"
        >
          All Products
        </a>
      </div>
      <div
        class="sc-htpNat KtFsv"
      >
        <a
          class="sc-bxivhb cHbCFb"
          href="/cart"
        >
          My Cart
        </a>
        |
        <a
          class="sc-bxivhb cHbCFb"
          href="/login"
        >
          Log In
        </a>
         |
        <a
          class="sc-bxivhb cHbCFb"
          href="/signup"
        >
          Sign Up
        </a>
      </div>
    </header>
    <div
      class="sc-bdVaJa eGntFc"
    >
      <h1
        class="sc-htoDjs gUiwOs"
      >
        Your cart
      </h1>
      <ul />
    </div>
  </div>
  .c2 {
  padding: 2rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-bottom: 0.1rem solid gainsboro;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 3rem;
}

.c1 {
  margin: 0 1rem;
}

.c3 {
  font-family: circular,serif;
  font-weight: 400;
  color: #ef0d33;
  text-transform: uppercase;
}

.c4 {
  background: #ef0d33;
  cursor: pointer;
  padding: 1rem;
  margin-top: 0.5rem;
  border: none;
  border-radius: 5px;
  color: #fff;
}

<div>
    <header
      class="c0"
    >
      <div
        class=""
      >
        <a
          class="c1"
          href="/products"
        >
          All Products
        </a>
      </div>
      <div
        class=""
      >
        <a
          class="c1"
          href="/cart"
        >
          My Cart
        </a>
        |
        <a
          class="c1"
          href="/login"
        >
          Log In
        </a>
         |
        <a
          class="c1"
          href="/signup"
        >
          Sign Up
        </a>
      </div>
    </header>
    <div
      class="c2"
    >
      <h1
        class="c3"
      >
        Your cart
      </h1>
      <ul>
        <li>
          <p>
            First Test Product
             - 
            1
          </p>
          <button
            class="c4"
            type="button"
          >
            Remove
          </button>
        </li>
      </ul>
    </div>
  </div>
</body>
`;
